solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/MetricsFetcher.java

```diff
@@ -0,0 +1,202 @@
+            Number n = (Number) node.get("CONTAINER.cores." + leafCoreMetricName);
+            if (n != null) count += n.intValue();
+          }
+          ctx.tags.put(CORES, count);
+        }
+        if (requestedTags.contains(SYSLOADAVG)) {
+          Number n = (Number) Utils.getObjectByPath(metrics, true, "solr.jvm/os.systemLoadAverage");
+          if (n != null) ctx.tags.put(SYSLOADAVG, n.doubleValue());
+        }
+        if (requestedTags.contains(HEAPUSAGE)) {
+          Number n = (Number) Utils.getObjectByPath(metrics, true, "solr.jvm/memory.heap.usage");
+          if (n != null) ctx.tags.put(HEAPUSAGE, n.doubleValue() * 100.0d);
+        }
+      }
+    } catch (Exception e) {
+      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Error getting remote info", e);
+    }
+  }
+
+  public void getTags(
+      String solrNode, Set<String> requestedTags, SolrClientNodeStateProvider.RemoteCallCtx ctx) {
+    try {
+      if (requestedTags.contains(NODE)) ctx.tags.put(NODE, solrNode);
+      if (requestedTags.contains(HOST)) {
+        Matcher hostAndPortMatcher = hostAndPortPattern.matcher(solrNode);
+        if (hostAndPortMatcher.find()) ctx.tags.put(HOST, hostAndPortMatcher.group(1));
+      }
+      if (requestedTags.contains(PORT)) {
+        Matcher hostAndPortMatcher = hostAndPortPattern.matcher(solrNode);
+        if (hostAndPortMatcher.find()) ctx.tags.put(PORT, hostAndPortMatcher.group(2));
+      }
+      if (requestedTags.contains(ROLE)) fillRole(solrNode, ctx, ROLE);
+      if (requestedTags.contains(NODEROLE))
+        fillRole(solrNode, ctx, NODEROLE); // for new policy framework
+
+      getRemoteInfo(solrNode, requestedTags, ctx);
+    } catch (Exception e) {
+      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, e);
+    }
+  }
+
+  private void fillRole(String solrNode, SolrClientNodeStateProvider.RemoteCallCtx ctx, String key)
+      throws KeeperException, InterruptedException {
+    Map<?, ?> roles =
+        (Map<?, ?>)
+            (ctx.session != null
+                ? ctx.session.get(ZkStateReader.ROLES)
+                : null); // we don't want to hit the ZK for each node
+    // so cache and reuse
+    try {
+      if (roles == null) roles = ctx.getZkJson(ZkStateReader.ROLES);
+      cacheRoles(solrNode, ctx, key, roles);
+    } catch (KeeperException.NoNodeException e) {
+      cacheRoles(solrNode, ctx, key, Collections.emptyMap());
+    }
+  }
+
+  private void cacheRoles(
+      String solrNode, SolrClientNodeStateProvider.RemoteCallCtx ctx, String key, Map<?, ?> roles) {
+    if (ctx.session != null) ctx.session.put(ZkStateReader.ROLES, roles);
+    if (roles != null) {
+      for (Map.Entry<?, ?> e : roles.entrySet()) {
+        if (e.getValue() instanceof List) {
+          if (((List<?>) e.getValue()).contains(solrNode)) {
+            ctx.tags.put(key, e.getKey());
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  public enum Variable {
```

Comment:
At least a one line javadoc on what this is.  Seems similar to the other "tags" although the value varies?  Say that.